src/i2c_software.c

```diff
@@ -0,0 +1,164 @@
+// Software I2C emulation
+//
+// Copyright (C) 2023  Kevin O'Connor <kevin@koconnor.net>
+// Copyright (C) 2023  Alan.Ma <tech@biqu3d.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <string.h> // memcpy
+#include "board/gpio.h" // gpio_out_setup
+#include "board/internal.h" // gpio_peripheral
+#include "board/misc.h" // timer_read_time
+#include "basecmd.h" // oid_alloc
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // sched_shutdown
+#include "i2ccmds.h" // i2cdev_set_software_bus
+
+struct i2c_software {
+    struct gpio_out scl, sda;
+    uint32_t sda_pin;
+    uint8_t addr;
+    unsigned int ticks;
+};
+
+void
+command_i2c_set_software_bus(uint32_t *args)
+{
+    struct i2cdev_s *i2c = i2cdev_oid_lookup(args[0]);
+    struct i2c_software *is = alloc_chunk(sizeof(*is));
+    // Max speed = 1MHz
+    unsigned int speed = args[3] > 1000000 ? 1000000 : args[3];
+    is->ticks = 1000000000 / speed / 2;
```

Comment:
Many MCUs don't implement a divide instruction.  This change would cause gcc to add software divide code to all MCUs that enable i2c support (eg, the AVR chips).  Is it necessary to support arbitrary speeds?  If only a speed of 100000 is needed then the code can be optimized so that the divide is done at compile time.

Reply:
Fixed speed to 100Khz https://github.com/Klipper3d/klipper/pull/6141/commits/c80f4a1244e30845aa2a791adbcb679560aadd32, avoid division during runtime.